### ISSUE TITLE
Implement partial application

### ISSUE THREAD

**Issue by PaulBone** [PlasmaLang/plasma#71]
Implement lambda expressions and partial application, including of higher order terms.

Lambda expressions should look like

    func(x, y) { return x + y }

Or with a shorthand when the body is a single expression.

    func(x, y) => x + y

Type declarations arn't shown here, they won't normally be used, but probably will be in the initial version.

Depends on #52

**Comment by PaulBone** [comment]
Duplicate of #164 and #165
